src/combobox/ComboBox.ts

```diff
@@ -0,0 +1,426 @@
+		});
+	}
+
+	private _restoreFocus() {
+		const func = this._focused ? 'focus' : 'blur';
+		this._inputElement[func]();
+	}
+
+	private _selectResult(result: any) {
+		const {
+			autoBlur,
+			onChange
+		} = this.properties;
+
+		this.updateSelectedIndex(Operation.reset);
+		this._open = false;
+		this._focused = !autoBlur;
+		this._ignoreFocus = true;
+		onChange && onChange(this._getResultLabel(result));
+	}
+
+	private _skipResult() {
+		this.updateSelectedIndex(this._direction);
+		this.invalidate();
+	}
+
+	@onPropertiesChanged
+	protected onPropertiesChanged(evt: PropertiesChangeEvent<this, ComboBoxProperties>) {
+		if (!this.registry || includes(evt.changedPropertyKeys, 'customResultItem') || includes(evt.changedPropertyKeys, 'customResultMenu')) {
+			this.registry = this._createRegistry();
+		}
+	}
+
+	protected scrollIntoView(element: HTMLElement) {
+		const menu = <HTMLElement> element.parentElement;
+		// Scroll menu up so top of highlighted result aligns with top of menu container
+		if (element.offsetTop - menu.scrollTop < 0) {
+			menu.scrollTop = element.offsetTop;
+		}
+		// Scroll menu down so bottom of highlighted result aligns with bottom of menu container
+		else if ((element.offsetTop - menu.scrollTop + element.offsetHeight) > menu.clientHeight) {
+			menu.scrollTop = element.offsetTop - menu.clientHeight + element.offsetHeight;
+		}
+	}
+
+	protected updateSelectedIndex(operation?: Operation, index?: number) {
+		const { results } = this.properties;
+
+		if (!results || results.length === 0 || !this._open) {
+			return;
+		}
+
+		if (index !== undefined) {
+			this._selectedIndex = index;
+		}
+		else if (operation !== undefined) {
+			if (operation === Operation.increase || operation === Operation.decrease) {
+				const total = results.length;
+				const base = operation === Operation.increase ? 0 : -1;
+				const current = this._selectedIndex !== undefined ? this._selectedIndex + operation : base;
+				this._selectedIndex = (current + total) % total;
+			}
+			else {
+				this._selectedIndex = undefined;
+			}
+			this._direction = operation;
+		}
+	}
+
+	render(): DNode {
+		const {
+			clearable,
+			disabled,
+			required,
+			invalid,
+			formId,
+			label,
+			readOnly,
+			results = [],
+			inputProperties = {},
+			value = ''
+		} = this.properties;
+
+		const menu = this._renderMenu(results);
+
+		this._notifyMenuChange();
+
+		this._wasOpen = this._open;
+
+		const container = v('div', {
+			bind: this,
+			classes: this.classes(css.combobox),
+			afterCreate: this._afterCreate,
+			afterUpdate: this._afterUpdate
+		}, [
+			// Use TextInput once landed
+			w(TextInput, {
+				bind: this,
+				classes: this.classes(clearable ? css.clearable : null),
+				disabled,
+				readOnly,
+				invalid,
+				required,
+				onBlur: this._onInputBlur,
+				onFocus: this._onInputFocus,
+				onInput: this._onInput,
+				onKeyDown: this._onInputKeyDown,
+				value,
+				overrideClasses: css,
+				...inputProperties
```

Comment:
The input should have `aria-controls` set to the id of the listbox, but right now the widget doesn't support adding arbitrary aria properties. I'm not entirely sure if it should just have a way to set random props, or if we should just add `aria-controls`

I think it'd also be better to set the label on the TextInput element, instead of wrapping the entire combobox, since W3C recommends setting the label's `for` attribute to the text input's id